src/aleph/http/client.clj

```diff
@@ -692,33 +701,35 @@
             (cond
 
               (not (.isHandshakeComplete handshaker))
-              (when (.cancel ^Future @timeout-task false)
-                (-> (netty/wrap-future (.processHandshake handshaker ch msg))
-                    ;; we want to check timeout here too
-                    (d/timeout! handshake-timeout)
-                    (d/chain'
-                     (fn [_]
-                       (let [out (netty/sink ch false
-                                             (http/websocket-message-coerce-fn ch pending-pings)
-                                             (fn [] @desc))]
-
-                         (s/on-closed out (fn [] (http/resolve-pings! pending-pings false)))
-
-                         (d/success! d
-                                     (doto
-                                         (s/splice out @in)
-                                       (reset-meta! {:aleph/channel ch})))
-
-                         (s/on-drained @in
-                                       #(when (.isOpen ch)
-                                          (d/chain'
-                                           (netty/wrap-future (.close handshaker ch (CloseWebSocketFrame.)))
-                                           (fn [_] (netty/close ctx))))))))
-                    (d/catch'
-                        (fn [ex]
-                          ;; handle handshake exception
-                          (d/error! d ex)
-                          (netty/close ctx)))))
+              (-> (.processHandshake handshaker ch msg)
+                  netty/wrap-future
+                  (d/chain'
+                   (fn [_]
+                     ;; cancel timeout if it was scheduled
+                     (when-let [^ScheduledFuture timeout-future @timeout-task]
+                       (.cancel timeout-future false)
```

Comment:
Isn't there a race condition ? Client can get handshake completed and timeout event can get triggered on different thread ?

> Avoid `.cancel` sync on I/O threads

Well, that is `.cancel` call again. I don't think it will cause any performance issues though - it is single CAS call to mark future as cancelled

Reply:
It's a future listener, it executes not on I/O thread

Reply:
We rely on the fact that `(.isHandshakeComplete handshaker)` inside timeout callback returns `false` if  `handshakeFuture` realized.